src/tty.rs

```diff
@@ -201,92 +178,73 @@ fn get_pw_entry(buf: &mut [i8; 1024]) -> Passwd {
-            Some(shell) => shell,
-            None => die!("Invalid shell value")
-        },
-        None => pw.shell
-    };
-
-    // setup environment
-    env::set_var("LOGNAME", pw.name);
-    env::set_var("USER", pw.name);
-    env::set_var("SHELL", shell);
-    env::set_var("HOME", pw.dir);
-    env::set_var("TERM", "xterm-256color"); // sigh
-
-    unsafe {
-        libc::signal(libc::SIGCHLD, libc::SIG_DFL);
-        libc::signal(libc::SIGHUP, libc::SIG_DFL);
-        libc::signal(libc::SIGINT, libc::SIG_DFL);
-        libc::signal(libc::SIGQUIT, libc::SIG_DFL);
-        libc::signal(libc::SIGTERM, libc::SIG_DFL);
-        libc::signal(libc::SIGALRM, libc::SIG_DFL);
-    }
-
-    // pw.shell is null terminated
-    let shell = unsafe { CStr::from_ptr(shell.as_ptr() as *const _) };
-
-    let argv = [shell.as_ptr(), ptr::null()];
-
-    let res = unsafe {
-        libc::execvp(shell.as_ptr(), argv.as_ptr())
-    };
+    let (master, slave) = openpty(win.ws_row as _, win.ws_col as _);
+    let mut shell = OsString::from(pw.shell);
 
-    if res < 0 {
-        die!("execvp failed: {}", errno());
+    if let Some(config_shell) = config.shell() {
+        shell = config_shell;
     }
 
-    ::std::process::exit(1);
-}
+    let mut builder = process::Command::new(shell.as_os_str());
 
-/// Create a new tty and return a handle to interact with it.
-pub fn new<T: ToWinsize>(config: &Config, size: T) -> Pty {
-    let win = size.to_winsize();
-
-    let (master, slave) = openpty(win.ws_row as _, win.ws_col as _);
+    // Setup child stdin/stdout/stderr as slave fd of pty
+    builder.stdin(unsafe { process::Stdio::from_raw_fd(slave) });
+    builder.stderr(unsafe { process::Stdio::from_raw_fd(slave) });
+    builder.stdout(unsafe { process::Stdio::from_raw_fd(slave) });
 
-    match fork() {
-        Relation::Child => {
-            unsafe {
-                // Create a new process group
-                libc::setsid();
+    // Setup environment
+    builder.env("LOGNAME", pw.name);
+    builder.env("USER", pw.name);
+    builder.env("SHELL", shell);
+    builder.env("HOME", pw.dir);
+    builder.env("TERM", "xterm-256color"); // sigh
 
-                // Duplicate pty slave to be child stdin, stdoud, and stderr
-                libc::dup2(slave, 0);
-                libc::dup2(slave, 1);
-                libc::dup2(slave, 2);
+    builder.before_exec(move || {
+        // Create a new process group
+        unsafe {
+            let err = libc::setsid();
+            if err == -1 {
+                die!("Failed to set session id: {}", errno());
             }
+        }
 
-            set_controlling_terminal(slave);
+        set_controlling_terminal(slave);
 
-            // No longer need slave/master fds
-            unsafe {
-                libc::close(slave);
-                libc::close(master);
-            }
+        //// No longer need slave/master fds
+        //unsafe {
+        //    libc::close(slave);
```

Comment:
Yeah, the borrowchecker don't like that I use slave/master in both child and parent..